http-response/background.js

```diff
@@ -1,22 +1,154 @@
 function listener(details) {
+  // The received data is a stream of bytes. In order to do text-based
+  // modifications, it is necessary to decode the bytes into a string
+  // using the proper character encoding, do any modifications, then
+  // encode back into a stream of bytes.
+  // Historically, detecting character encoding has been a tricky task
+  // taken on by the browser. Here, a simplified approach is taken
+  // and the complexity is hidden in a helper method.
+  let decoder, encoder;
+  [decoder, encoder] = detectCharsetAndSetupDecoderEncoder(details);
   let filter = browser.webRequest.filterResponseData(details.requestId);
-  let decoder = new TextDecoder("utf-8");
-  let encoder = new TextEncoder();
+  let fullStr = '';
+  
+  filter.ondata = e => {
+    // Note that the event's data may break in the middle of an encoded
+    // character - the stream parameter is critical for success as this
+    // method gets called multiple times.
+    let str = decoder.decode(e.data, {stream: true});
+    fullStr += str;
+  }
+  
+  filter.onstop = async e => {
+    fullStr += decoder.decode(); //Flush the buffer
+    // Just change any instance of Example or Test in the HTTP response
+    // to WebExtension Example or WebExtension Test.
+    let mutatedStr = fullStr.replace(/Example/g, 'WebExtension Example');
+    mutatedStr = mutatedStr.replace(/Test/g, 'WebExtension Test');
+    filter.write(encoder.encode(mutatedStr));
+    filter.close();
+  }
 
-  filter.ondata = event => {
-    let str = decoder.decode(event.data, {stream: true});
-    // Just change any instance of Example in the HTTP response
-    // to WebExtension Example.
-    str = str.replace(/Example/g, 'WebExtension Example');
-    filter.write(encoder.encode(str));
-    filter.disconnect();
+  filter.onerror = e => {
+    try {
+        filter.close();

```

Comment:
Why do you want to explicitly handle errors here?

`filter.onerror` could occur when the response is redirected (when you've failed to check for status codes, as you did before),
and also repeatedly when an out-of-memory issue occurs - https://bugzilla.mozilla.org/show_bug.cgi?id=1404381

Reply:
Good question - this part of the error handling isn't particularly relevant to the heart of the example either.

- [ ] Remove explicit error handling